sbncode/EventGenerator/MultiPart/MultiPartRain_module.cc

```diff
@@ -374,46 +412,57 @@ void MultiPartRain::GenMomentum(const PartGenParam& param, const double& mass, d
       }
 
       while(1) {
-	double phi   = fFlatRandom->fire(0, 2 * 3.141592653589793238);
-	double theta = fFlatRandom->fire(0, 1 * 3.141592653589793238);
-	px = cos(phi) * sin(theta);
-	py = sin(phi) * sin(theta);
-	pz = cos(theta);
-	if( (px * sign_x) >= 0. && 
-	    (py * sign_y) >= 0. &&
-	    (pz * sign_z) >= 0. )
-	  break;
+
+				if (_cosmic_distribution) {
+					double phi   = fFlatRandom->fire(0, 2 * 3.141592653589793238);
+					double theta = fGeneralRandom->fire() * 3.141592653589793238;
+					pz = cos(phi) * sin(theta);
+					px = sin(phi) * sin(theta);
+					py = cos(theta);
+				} else {
+					px = fNormalRandom->fire(0, 1);
+					py = fNormalRandom->fire(0, 1);
+					pz = fNormalRandom->fire(0, 1);
+					double p = std::sqrt( std::pow(px, 2) + std::pow(py, 2) + std::pow(pz, 2) );
+					px = px / p;
+					py = py / p;
+					pz = pz / p;
+				}
+
+				if( (px * sign_x) >= 0. &&
+						(py * sign_y) >= 0. &&
+						(pz * sign_z) >= 0. )
```

Comment:
Currently all distributions are symmetric w.r.t. zero in all coordinates, yes, so that would probably be fine... but in case we ever one day decide to add a distribution that does not respect this assumption, I will keep this as is.